File: producer/order_producer.py
"""
Fake order event producer — generates realistic e-commerce orders and
publishes them to a Kafka topic at a configurable rate.

Demonstrates:
  - Idempotent producer (enable.idempotence=True)  → exactly-once delivery
  - Key-based partitioning (customer_id as key)     → related events in order
  - Snappy compression                              → ~3x smaller on the wire
  - AdminClient topic auto-creation
"""
import argparse
import json
import logging
import random
import sys
import time
import uuid
from datetime import datetime, timezone

from confluent_kafka import Producer
from confluent_kafka.admin import AdminClient, NewTopic

logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s %(levelname)s %(message)s",
    datefmt="%H:%M:%S",
)
logger = logging.getLogger("producer")

PRODUCTS = {
    "prod-001": 9.99,
    "prod-002": 24.99,
    "prod-003": 49.99,
    "prod-004": 99.99,
    "prod-005": 199.99,
}
STATUSES = ["PLACED", "CONFIRMED", "SHIPPED", "DELIVERED", "CANCELLED"]


# ── topic helpers ──────────────────────────────────────────────────────────────

def ensure_topic(servers: str, topic: str, partitions: int = 3):
    admin = AdminClient({"bootstrap.servers": servers})
    existing = admin.list_topics(timeout=10).topics
    if topic in existing:
        logger.info("Topic '%s' already exists (%d partitions)", topic, partitions)
        return
    logger.info("Creating topic '%s' with %d partitions…", topic, partitions)
    futures = admin.create_topics(
        [NewTopic(topic, num_partitions=partitions, replication_factor=1)]
    )
    for t, fut in futures.items():
        try:
            fut.result()
            logger.info("Topic '%s' created", t)
        except Exception as exc:
            logger.warning("Could not create topic '%s': %s", t, exc)


# ── order factory ──────────────────────────────────────────────────────────────

def make_order() -> dict:
    product_id = random.choice(list(PRODUCTS))
    quantity = random.randint(1, 5)
    return {
        "order_id":    str(uuid.uuid4()),
        "customer_id": f"cust-{random.randint(1000, 9999)}",
        "product_id":  product_id,
        "quantity":    quantity,
        "price":       round(PRODUCTS[product_id] * quantity, 2),
        "status":      random.choice(STATUSES),
        "event_time":  datetime.now(timezone.utc).isoformat(),
    }


# ── delivery callback ──────────────────────────────────────────────────────────

def _on_delivery(err, msg):
    if err:
        logger.error("Delivery failed for offset %s: %s", msg.offset(), err)


# ── main ───────────────────────────────────────────────────────────────────────

def main():
    p = argparse.ArgumentParser(description="StreamLake order event producer")
    p.add_argument("--servers", default="localhost:9092", help="Kafka bootstrap servers")
    p.add_argument("--topic",   default="orders",         help="Target topic name")
    p.add_argument("--rate",    type=float, default=2.0,  help="Events per second")
    p.add_argument("--count",   type=int,   default=0,    help="Total events (0 = infinite)")
    args = p.parse_args()

    ensure_topic(args.servers, args.topic)

    producer = Producer({
        "bootstrap.servers":     args.servers,
        "acks":                  "all",        # wait for all replicas
        "enable.idempotence":    True,         # exactly-once to the broker
        "compression.type":      "snappy",
        "linger.ms":             5,            # micro-batching
        "batch.size":            65536,
    })

    interval = 1.0 / args.rate
    sent = 0

    logger.info(
        "Producing to topic='%s' at %.1f events/sec  (Ctrl-C to stop)",
        args.topic, args.rate,
    )

    try:
        while args.count == 0 or sent < args.count:
            order = make_order()
            producer.produce(
                topic=args.topic,
                key=order["customer_id"].encode(),
                value=json.dumps(order).encode(),
                callback=_on_delivery,
            )
            producer.poll(0)  # trigger delivery callbacks without blocking
            sent += 1
            if sent % 100 == 0:
                logger.info("Produced %d events", sent)
            time.sleep(interval)

    except KeyboardInterrupt:
        logger.info("\nStopping — flushing remaining messages…")

    finally:
        producer.flush(timeout=15)
        logger.info("Done. Total produced: %d", sent)


if __name__ == "__main__":
    main()

File: data-plane/src/main/java/com/streamlake/producer/OrderProducer.java
package com.streamlake.producer;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.streamlake.model.Order;
import org.apache.kafka.clients.admin.AdminClient;
import org.apache.kafka.clients.admin.AdminClientConfig;
import org.apache.kafka.clients.admin.NewTopic;
import org.apache.kafka.clients.producer.*;
import org.apache.kafka.common.serialization.StringSerializer;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.time.Instant;
import java.util.*;
import java.util.concurrent.ThreadLocalRandom;

/**
 * Idempotent Kafka producer that generates fake e-commerce order events.
 *
 * Key Confluent patterns demonstrated:
 *  - enable.idempotence=true       → exactly-once to the broker (PID + sequence no.)
 *  - acks=all                      → waits for all in-sync replicas
 *  - compression.type=snappy       → ~3x smaller on the wire
 *  - linger.ms + batch.size        → micro-batching for throughput
 *  - customer_id as message key    → guarantees per-customer ordering
 */
public class OrderProducer {

    private static final Logger log = LoggerFactory.getLogger(OrderProducer.class);
    private static final ObjectMapper MAPPER = new ObjectMapper();

    private static final String[] PRODUCTS = {"prod-001","prod-002","prod-003","prod-004","prod-005"};
    private static final double[] PRICES   = {9.99, 24.99, 49.99, 99.99, 199.99};
    private static final String[] STATUSES = {"PLACED","CONFIRMED","SHIPPED","DELIVERED","CANCELLED"};

    private final KafkaProducer<String, String> producer;
    private final String topic;
    private final String bootstrapServers;

    public OrderProducer(String bootstrapServers, String topic) {
        this.topic = topic;
        this.bootstrapServers = bootstrapServers;

        Properties props = new Properties();
        props.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG,  bootstrapServers);
        props.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG,   StringSerializer.class.getName());
        props.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, StringSerializer.class.getName());
        props.put(ProducerConfig.ACKS_CONFIG,               "all");
        props.put(ProducerConfig.ENABLE_IDEMPOTENCE_CONFIG, true);
        props.put(ProducerConfig.COMPRESSION_TYPE_CONFIG,   "snappy");
        props.put(ProducerConfig.LINGER_MS_CONFIG,          5);
        props.put(ProducerConfig.BATCH_SIZE_CONFIG,         65536);
        props.put(ProducerConfig.RETRIES_CONFIG,            Integer.MAX_VALUE);
        props.put(ProducerConfig.MAX_IN_FLIGHT_REQUESTS_PER_CONNECTION, 5);

        this.producer = new KafkaProducer<>(props);
    }

    public void ensureTopic(int partitions) {
        Properties adminProps = new Properties();
        adminProps.put(AdminClientConfig.BOOTSTRAP_SERVERS_CONFIG, bootstrapServers);

        try (AdminClient admin = AdminClient.create(adminProps)) {
            Set<String> existing = admin.listTopics().names().get();
            if (!existing.contains(topic)) {
                admin.createTopics(List.of(new NewTopic(topic, partitions, (short) 1))).all().get();
                log.info("Created topic '{}' ({} partitions)", topic, partitions);
            } else {
                log.info("Topic '{}' already exists", topic);
            }
        } catch (Exception e) {
            log.warn("Could not ensure topic: {}", e.getMessage());
        }
    }

    public Order generateOrder() {
        ThreadLocalRandom rng = ThreadLocalRandom.current();
        int idx = rng.nextInt(PRODUCTS.length);
        int qty = rng.nextInt(1, 6);

        Order o = new Order();
        o.setOrderId(UUID.randomUUID().toString());
        o.setCustomerId("cust-" + rng.nextInt(1000, 9999));
        o.setProductId(PRODUCTS[idx]);
        o.setQuantity(qty);
        o.setPrice(Math.round(PRICES[idx] * qty * 100.0) / 100.0);
        o.setStatus(STATUSES[rng.nextInt(STATUSES.length)]);
        o.setEventTime(Instant.now().toString());
        return o;
    }

    public void produce(double ratePerSec, long totalCount) throws Exception {
        ensureTopic(3);
        long intervalMs = Math.max(1, (long) (1000.0 / ratePerSec));
        long sent = 0;

        log.info("Producing to '{}' at {}/sec  (Ctrl-C to stop)", topic, ratePerSec);
        try {
            while (totalCount == 0 || sent < totalCount) {
                Order order = generateOrder();
                String json  = MAPPER.writeValueAsString(order);

                producer.send(
                    new ProducerRecord<>(topic, order.getCustomerId(), json),
                    (meta, err) -> {
                        if (err != null) log.error("Delivery failed: {}", err.getMessage());
                    }
                );
                if (++sent % 100 == 0) log.info("Produced {} events", sent);
                Thread.sleep(intervalMs);
            }
        } finally {
            producer.flush();
            producer.close();
            log.info("Producer stopped. Total sent: {}", sent);
        }
    }
}
